woocommerce/compatibility/class-sv-wc-order-compatibility.php

```diff
@@ -522,6 +522,121 @@ public static function get_edit_order_url( \WC_Order $order ) {
 	}
 
 
+	/**
+	 * Gets the order meta according to HPOS availability.
+	 * Uses WC_Order data store object if HPOS is enabled, otherwise it uses the traditional get_post_meta() method.
+	 *
+	 * @since x.y.z
+	 *
+	 * @param int $order_id Required. Order ID.
+	 * @param string $key Required. Meta key.
+     * @param bool $single return first found meta with key, or all with $key.
+	 * 			   Defaults to true if HPOS is enabled, otherwise false.
+     *
+	 * @return mixed
+	 */
+	public static function get_order_meta( int $order_id, string $key, $single = null) {
+
+		$is_hpos_enabled = SV_WC_Plugin_Compatibility::is_hpos_enabled();
+		$value = false;
+
+		// setting $single to false by default if hpos is disabled to match it with get_post_meta method's default value
+		if( is_null($single) ) {
+			$single = $is_hpos_enabled;
+		}
+
+		if ( $is_hpos_enabled ) {
+			$order = wc_get_order($order_id);
+			if( $order ){
+				$value = $order->get_meta($key, $single);
+			}
+		} else {
+			$value = get_post_meta( $order_id, $key, $single );
+		}
+
+		return $value;
+	}
+
+	/**
+	 * Updates the order meta according to HPOS availability.
+	 * Uses WC_Order data store object if HPOS is enabled, otherwise it uses the traditional update_post_meta() method.
+	 *
+	 * @since x.y.z
+	 *
+	 * @param int $order_id Required. Order ID.
+	 * @param string $key Required. Meta key.
+     * @param mixed $value Required. Meta value.
+     *
+	 */
+	public static function update_order_meta( int $order_id, string $key, $value ) {
+
+		$is_hpos_enabled = SV_WC_Plugin_Compatibility::is_hpos_enabled();
+
+		if ( $is_hpos_enabled ) {
+			$order = wc_get_order($order_id);
+			if( $order ){
+				$order->update_meta_data($key, $value, $order_id);
+				$order->save();
+			}
+		} else {
+			update_post_meta( $order_id, $key, $value );
+		}
+	}
+
+
+	/**
+	 * Adds the order meta according to HPOS availability.
+	 * Uses WC_Order data store object if HPOS is enabled, otherwise it uses the traditional add_post_meta() method.
+	 *
+	 * @since x.y.z
+	 *
+	 * @param int $order_id Required. Order ID.
+	 * @param string $key Required. Meta key.
+     * @param mixed $value Required. meta value.
+	 * @param bool $unique Optional. Whether the same key should not be added.
+	 * 			   Default false.
+     *
+	 */
+	public static function add_order_meta( int $order_id, string $key, $value, $unique = false ) {
+
+		$is_hpos_enabled = SV_WC_Plugin_Compatibility::is_hpos_enabled();
+
+		if ( $is_hpos_enabled ) {
+			$order = wc_get_order($order_id);
+			if( $order ){
+				$order->add_meta_data( $key, $value, $unique );
+				$order->save();
```

Comment:
I don't think we need to save the whole order

save_meta_data() would suffice

Reply:
@unfulvio-godaddy yes, this makes sense but I am following this HPOS guide [here](https://github.com/woocommerce/woocommerce/wiki/High-Performance-Order-Storage-Upgrade-Recipe-Book#apis-for-gettingsetting-posts-and-postmeta), so let me know if that's fine or should I replace it with save_meta_data()?

Reply:
hmmm, normally save_meta_data() is sufficient, but I wonder why they suggest that - while they also add

> Calling the save() method is a relatively expensive operation, so you may wish to avoid calling it more times than necessary (for example, if you know it will be called later in the same flow, you may wish to avoid additional earlier calls when operating on the same object).

so I wonder if there's something special under the hood when it comes to HPOS and `save()` is necessary

we should investigate if this assumption is correct, or really there's no difference, and save_meta_data is sufficient

if it's not, I would still not place `save()` here - because in the scenario one is setting multiple meta data at once, then this would become a very expensive operation (say one is handling meta in a foreach loop, then save should be called at the very end) - in that case we need a different strategy to have plugins call save when appropriate

Reply:
Upon suggestion of @jstern-godaddy I have contacted the author of that wiki page directly to ask for clarifications.

In the meantime could you leave this to `save_meta_data` and later experiment in QA/testing to see if there's any drawback going with that?

thank you

Reply:
so, it turns out this is due to a WC core bug: https://godaddy.slack.com/archives/C026JPZ4GLB/p1673409266503829

this won't halt further work with HPOS on our side however we would prefer to have that sorted before we deploy any plugins that include HPOS compatibility

I would say that for now we still use `save_meta_data` and we come back to this as soon there's a resolution from WC. We hold any FW release or plugin deploy before we are sure. It's also likely we will need an additional condition check here https://github.com/godaddy-wordpress/wc-plugin-framework/pull/582#discussion_r1066554703 before declaring compatibility

Reply:
Understood, thanks for doing the research and sharing these details. I am on it!